test/test_containers.py

```diff
@@ -24,9 +25,51 @@ def test_attrs_from(ss_container):
     assert len(ts.attrs) == 1
     assert ts.attrs["test_attr1"] == "hello"
     assert ts.vis.attrs["test_attr2"] == "hello2"
+    assert ts.weight.attrs["test_attr3"] == "hello3"
 
-    # Check that the axis attributes on the dataasets did not get overwritten
+    # Check that the axis attributes on the datasets did not get overwritten
     assert len(ts.vis.attrs) == 2
-    assert len(ts.weight.attrs) == 1
+    assert len(ts.weight.attrs) == 2
     assert tuple(ts.vis.attrs["axis"]) == ("freq", "stack", "time")
     assert tuple(ts.weight.attrs["axis"]) == ("freq", "stack", "time")
+
+
+def test_copy(ss_container):
+
+    ss_container.vis[:] = np.arange(16)
+    ss_container.weight[:] = np.arange(16)
+    ss_container.weight.attrs
+
+    ss_copy = ss_container.copy(shared=("vis",))
+
+    # Check that the copy has worked
+    # Check the dataset values...
+    assert (ss_copy.vis[:] == ss_container.vis[:]).all()
+    assert (ss_copy.weight[:] == ss_container.weight[:]).all()
+    # Check the attributes...
+    assert ss_copy.attrs["test_attr1"] == "hello"
+    assert ss_copy.vis.attrs["test_attr2"] == "hello2"
+    assert ss_copy.weight.attrs["test_attr3"] == "hello3"
+
+    # Modify the datasets
+    ss_container.vis[:] = 1.0
+    ss_container.weight[:] = 2.0
+    ss_container.vis.attrs["test_attr4"] = "hello4"
+    ss_container.weight.attrs["test_attr5"] = "hello5"
+
+    # Check that we see the modified values
+    assert (ss_copy.vis[:] == 1.0).all()
+    assert (ss_copy.weight[:] == np.arange(16)).all()
+    assert ss_copy.vis.attrs["test_attr4"] == "hello4"
+    assert "test_attr5" not in ss_copy.weight.attrs
+
+    # These tests only make sense when running across multiple MPI processes
+    assert ss_container.vis.distributed_axis == 0
+    assert ss_copy.vis.distributed_axis == 0
+
+    current_shape = ss_copy.weight.local_shape
+
+    ss_container.redistribute("ra")
```

Comment:
I hadn't realised on first pass that calling `redistribute` on either of the shared copies will lead to the other one having different distribution schemes between its datasets. This is tricky and could easily cause some funny behaviour if the user is not careful. It might be good to explicitly document this behaviour in `ContainerBase.copy` and maybe even in the tasks that make use of a `shared` config property.

Reply:
Yah. It's a little tricky I agree. You're right that we need to make sure that any class that does this clearly documents it, I tried that on the one task I've modified to do it, but it should maybe be even clearer.